sound/soc/codecs/rt5682-sdw.c

```diff
@@ -0,0 +1,591 @@
+	stream = snd_soc_dai_get_dma_data(dai, substream);
+
+	if (!stream)
+		return -ENOMEM;
+
+	if (!rt5682->slave)
+		return -EINVAL;
+
+	/* SoundWire specific configuration */
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+		direction = SDW_DATA_DIR_RX;
+		port = 1;
+	} else {
+		direction = SDW_DATA_DIR_TX;
+		port = 2;
+	}
+
+	stream_config.frame_rate = params_rate(params);
+	stream_config.ch_count = params_channels(params);
+	stream_config.bps = snd_pcm_format_width(params_format(params));
+	stream_config.direction = direction;
+
+	num_channels = params_channels(params);
+	port_config.ch_mask = (1 << (num_channels)) - 1;
+	port_config.num = port;
+
+	retval = sdw_stream_add_slave(rt5682->slave, &stream_config,
+				      &port_config, 1, stream->sdw_stream);
+	if (retval) {
+		dev_err(dai->dev, "Unable to configure port\n");
+		return retval;
+	}
+
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+		switch (params_rate(params)) {
+		case 48000:
+			val = RT5682_SDW_REF_1_48K;
+			break;
+		case 96000:
+			val = RT5682_SDW_REF_1_96K;
+			break;
+		case 192000:
+			val = RT5682_SDW_REF_1_192K;
+			break;
+		case 32000:
+			val = RT5682_SDW_REF_1_32K;
+			break;
+		case 24000:
+			val = RT5682_SDW_REF_1_24K;
+			break;
+		case 16000:
+			val = RT5682_SDW_REF_1_16K;
+			break;
+		case 12000:
+			val = RT5682_SDW_REF_1_12K;
+			break;
+		case 8000:
+			val = RT5682_SDW_REF_1_8K;
+			break;
+		case 44100:
+			val = RT5682_SDW_REF_1_44K;
+			break;
+		case 88200:
+			val = RT5682_SDW_REF_1_88K;
+			break;
+		case 176400:
+			val = RT5682_SDW_REF_1_176K;
+			break;
+		case 22050:
+			val = RT5682_SDW_REF_1_22K;
+			break;
+		case 11025:
+			val = RT5682_SDW_REF_1_11K;
+			break;
+		default:
+			return -EINVAL;
+		}
+
+		regmap_update_bits(rt5682->regmap, RT5682_SDW_REF_CLK,
+			RT5682_SDW_REF_1_MASK, val);
+
+		if (params_rate(params) <= 48000)
+			val = RT5682_DAC_OSR_D_8;
+		else if (params_rate(params) <= 96000)
+			val = RT5682_DAC_OSR_D_4;
+		else
+			val = RT5682_DAC_OSR_D_2;
+
+		regmap_update_bits(rt5682->regmap, RT5682_ADDA_CLK_1,
+			RT5682_DAC_OSR_MASK, val);
+	} else {
+		switch (params_rate(params)) {
+		case 48000:
+			val = RT5682_SDW_REF_2_48K;
+			break;
+		case 96000:
+			val = RT5682_SDW_REF_2_96K;
+			break;
+		case 192000:
+			val = RT5682_SDW_REF_2_192K;
+			break;
+		case 32000:
+			val = RT5682_SDW_REF_2_32K;
+			break;
+		case 24000:
+			val = RT5682_SDW_REF_2_24K;
+			break;
+		case 16000:
+			val = RT5682_SDW_REF_2_16K;
+			break;
+		case 12000:
+			val = RT5682_SDW_REF_2_12K;
+			break;
+		case 8000:
+			val = RT5682_SDW_REF_2_8K;
+			break;
+		case 44100:
+			val = RT5682_SDW_REF_2_44K;
+			break;
+		case 88200:
+			val = RT5682_SDW_REF_2_88K;
+			break;
+		case 176400:
+			val = RT5682_SDW_REF_2_176K;
+			break;
+		case 22050:
+			val = RT5682_SDW_REF_2_22K;
+			break;
+		case 11025:
+			val = RT5682_SDW_REF_2_11K;
+			break;
+		default:
+			return -EINVAL;
+		}
```

Comment:
to reduce the number of lines, you could have a val_p and val_c, set the right frequency and then decide to use val_p or val_c depending on  the direction?